### PR TITLE
Purview: Parallelize PSPC cold-cache scope refresh

### PR DESCRIPTION
**Pull request by taisirhassan**
### Motivation and Context

Improve Purview PSPC cold-cache behavior by allowing `ProcessContent` to run without waiting for a foreground `ProtectionScopes` lookup. This reduces user-visible latency on cache misses while still refreshing the `ProtectionScopes` cache for future requests.

The implementation aligns Agent Framework's Python and .NET Purview packages with the Purview SDK's parallel scope retrieval behavior.

### Description

- On a `ProtectionScopes` cache miss, refresh scopes in the background while `ProcessContent` runs in the foreground.
- Preserve cache-hit behavior: cached scopes are evaluated locally before deciding whether to process inline, queue offline processing, or send `ContentActivities`.
- Preserve SDK parity for inline evaluation: cold misses do not force `Prefer: evaluateInline`; cached inline scopes set inline evaluation explicitly.
- Add `.NET` `ScopeRetrievalJob` handling so background scope refreshes use the existing background job infrastructure.
- Send `ContentActivities` in the background when a cold-miss scope refresh finds no applicable scopes, preserving the audit signal.
- Cache tenant-level 402 (`Payment Required`) responses from background scope refreshes so subsequent requests for the tenant short-circuit before repeating Purview work.
- Invalidate cached scopes when `ProcessContent` reports modified protection-scope state.
- Preserve restriction-only policy actions and deduplicate combined policy actions by `(action, restriction_action)` in Python and .NET.
- Align .NET multi-location scope matching with Python and the Purview SDK by treating a scope as applicable when any location matches.
- Update the Python Purview sample to run the same `good -> expected block -> good` flow across agent middleware, chat middleware, custom cache, and default cache scenarios.

### Validation

- Python Purview processor tests: `36 passed`
- .NET Purview unit tests: `56 passed` on `net10.0`
- .NET Purview unit tests: `56 passed` on `net472`
- Python sample syntax check: `py_compile` passed
- Manual sample run verified the `good (cold cache) -> expected block -> good (warm cache)` orchestration, including cache miss/hit behavior and Purview prompt blocking with the configured `Prompt blocked by policy` message.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [x] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

### CODE REVIEW

**Comment by moonbox3** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/ce1779b5516107bbc0de89b872b80893a5aaa942/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/purview/agent_framework_purview</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/ce1779b5516107bbc0de89b872b80893a5aaa942/packages/purview/agent_framework_purview/_processor.py">_processor.py</a></td><td>184</td><td>12</td><td>93%</td><td><a href="https://github.com/microsoft/agent-framework/blob/ce1779b5516107bbc0de89b872b80893a5aaa942/packages/purview/agent_framework_purview/_processor.py#L178">178</a>, <a href="https://github.com/microsoft/agent-framework/blob/ce1779b5516107bbc0de89b872b80893a5aaa942/packages/purview/agent_framework_purview/_processor.py#L258-L261">258&ndash;261</a>, <a href="https://github.com/microsoft/agent-framework/blob/ce1779b5516107bbc0de89b872b80893a5aaa942/packages/purview/agent_framework_purview/_processor.py#L288">288</a>, <a href="https://github.com/microsoft/agent-framework/blob/ce1779b5516107bbc0de89b872b80893a5aaa942/packages/purview/agent_framework_purview/_processor.py#L316-L317">316&ndash;317</a>, <a href="https://github.com/microsoft/agent-framework/blob/ce1779b5516107bbc0de89b872b80893a5aaa942/packages/purview/agent_framework_purview/_processor.py#L328">328</a>, <a href="https://github.com/microsoft/agent-framework/blob/ce1779b5516107bbc0de89b872b80893a5aaa942/packages/purview/agent_framework_purview/_processor.py#L330">330</a>, <a href="https://github.com/microsoft/agent-framework/blob/ce1779b5516107bbc0de89b872b80893a5aaa942/packages/purview/agent_framework_purview/_processor.py#L336">336</a>, <a href="https://github.com/microsoft/agent-framework/blob/ce1779b5516107bbc0de89b872b80893a5aaa942/packages/purview/agent_framework_purview/_processor.py#L338">338</a></td></tr><tr><td><b>TOTAL</b></td><td><b>34131</b></td><td><b>3909</b></td><td><b>88%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 6801 | 30 :zzz: | 0 :x: | 0 :fire: | 1m 53s :stopwatch: |

File: python/packages/purview/agent_framework_purview/_processor.py [hunk 1]
```diff
@@ -231,18 +231,19 @@ async def _process_with_scopes(self, pc_request: ProcessContentRequest) -> Proce
...
-                # Cache the exception at tenant level so all subsequent requests for this tenant fail fast
-                await self._cache.set(tenant_payment_cache_key, ex, ttl_seconds=ttl_seconds)
-                raise
+            return await self._process_with_cached_scopes(pc_request, cached_ps_resp, cache_key)
 
+        task = asyncio.create_task(self._refresh_protection_scopes_background(ps_req, cache_key))
+        self._background_tasks.add(task)
+        task.add_done_callback(self._background_tasks.discard)
+        return await self._call_process_content(pc_request, cache_key, dlp_actions=[])
+
```
[the remark this answers is not in the document]
Comment (taisirhassan) [reply python/packages/purview/agent_framework_purview/_processor.py]: future concern

**Comment by taisirhassan** [comment]
> @taisirhassan please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="Microsoft"

File: dotnet/src/Microsoft.Agents.AI.Purview/BackgroundJobRunner.cs [hunk 2]
```diff
@@ -68,6 +71,10 @@ private async Task RunJobAsync(BackgroundJobBase job)
             case ContentActivityJob contentActivityJob:
                 _ = await this._purviewClient.SendContentActivitiesAsync(contentActivityJob.Request, CancellationToken.None).ConfigureAwait(false);
                 break;
+            case ScopeRetrievalJob scopeRetrievalJob:
+                var response = await this._purviewClient.GetProtectionScopesAsync(scopeRetrievalJob.Request, CancellationToken.None).ConfigureAwait(false);
+                await this._cacheProvider.SetAsync(scopeRetrievalJob.CacheKey, response, CancellationToken.None).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (taisirhassan) [reply dotnet/src/Microsoft.Agents.AI.Purview/BackgroundJobRunner.cs]: added a cache for the 402 state

File: dotnet/src/Microsoft.Agents.AI.Purview/BackgroundJobRunner.cs [hunk 3]
```diff
@@ -67,6 +71,21 @@ private async Task RunJobAsync(BackgroundJobBase job)
                 break;
             case ContentActivityJob contentActivityJob:
                 _ = await this._purviewClient.SendContentActivitiesAsync(contentActivityJob.Request, CancellationToken.None).ConfigureAwait(false);
+                break;
+            case ScopeRetrievalJob scopeRetrievalJob:
+                try
+                {
+                    var response = await this._purviewClient.GetProtectionScopesAsync(scopeRetrievalJob.Request, CancellationToken.None).ConfigureAwait(false);
```
Comment (Rishabh4275) [inline dotnet/src/Microsoft.Agents.AI.Purview/BackgroundJobRunner.cs]: Let's not use var

File: dotnet/src/Microsoft.Agents.AI.Purview/ScopedContentProcessor.cs [hunk 4]
```diff
@@ -193,22 +193,47 @@ private async Task<ProcessContentResponse> ProcessContentWithProtectionScopesAsy
...
         }
-        else
+        catch (PurviewJobException)
         {
-            psResponse = await this._purviewClient.GetProtectionScopesAsync(psRequest, cancellationToken).ConfigureAwait(false);
-            await this._cacheProvider.SetAsync(cacheKey, psResponse, cancellationToken).ConfigureAwait(false);
+            // QueueJob already logs failures. Scope warmup is best effort; don't block ProcessContent.
         }
 
+        return await this.CallProcessContentAsync(pcRequest, cacheKey, dlpActions: null, cancellationToken).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (taisirhassan) [reply dotnet/src/Microsoft.Agents.AI.Purview/ScopedContentProcessor.cs:224]: addressed

File: dotnet/src/Microsoft.Agents.AI.Purview/BackgroundJobRunner.cs (hunk 3 above)
Comment (taisirhassan) [reply dotnet/src/Microsoft.Agents.AI.Purview/BackgroundJobRunner.cs]: fixed to use the explicit type of `ProtectionScopesResponse`

**Comment by Rishabh4275** [comment]
1. Restore dedup in _combine_policy_actions (key by (action, restriction_action) or similar) and add a regression test for the duplicate-action case.
 2. Decide and document/align the 402-on-first-call contract between Python and .NET.

File: dotnet/src/Microsoft.Agents.AI.Purview/ScopedContentProcessor.cs [hunk 5]
```diff
@@ -193,43 +193,62 @@ private async Task<ProcessContentResponse> ProcessContentWithProtectionScopesAsy
...
+        try
+        {
+            this._channelHandler.QueueJob(new ScopeRetrievalJob(psRequest, cacheKey));
+        }
+        catch (PurviewJobException)
         {
-            psResponse = await this._purviewClient.GetProtectionScopesAsync(psRequest, cancellationToken).ConfigureAwait(false);
-            await this._cacheProvider.SetAsync(cacheKey, psResponse, cancellationToken).ConfigureAwait(false);
+            // QueueJob already logs failures. Scope warmup is best effort; don't block ProcessContent.
         }
```
[the remark this answers is not in the document]
Comment (taisirhassan) [reply dotnet/src/Microsoft.Agents.AI.Purview/ScopedContentProcessor.cs:223]: This is intentionally not set on the cold-cache path to match the Purview SDK parallel scope retrieval behavior.
The SDK queues scope retrieval in the background and calls ProcessContent foreground without forcing ProcessInline;
ProcessInline is only set on the cached-scope path when cached applicable scopes have ExecutionMode.EvaluateInline.
Forcing it here would diverge from SDK behavior.

**Comment by Rishabh4275** [comment]
Could we also check the cache miss contentActivities call path?

File: python/samples/05-end-to-end/purview_agent/sample_purview_agent.py [hunk 6]
```diff
@@ -138,14 +138,14 @@ def build_credential() -> Any:
 
 
 async def run_with_agent_middleware() -> None:
-    endpoint = os.environ.get("AZURE_OPENAI_ENDPOINT")
+    endpoint = os.environ.get("FOUNDRY_PROJECT_ENDPOINT")
     if not endpoint:
-        print("Skipping run: AZURE_OPENAI_ENDPOINT not set")
+        print("Skipping run: FOUNDRY_PROJECT_ENDPOINT not set")
```
[the remark this answers is not in the document]
Comment (Rishabh4275) [reply python/samples/05-end-to-end/purview_agent/sample_purview_agent.py:174]: You'll have to update the README as well accordingly

**Comment by Rishabh4275** [comment]
Check the README(4, all purview, one sample, and one with the actual package(python and .net)) as well. 
Also want to confirm that we're checking the good vs block path vs good path. 
Orchestrate multiple messages one after another.

File: dotnet/src/Microsoft.Agents.AI.Purview/BackgroundJobRunner.cs [hunk 7]
```diff
@@ -68,6 +75,43 @@ private async Task RunJobAsync(BackgroundJobBase job)
...
+                        new PaymentRequiredCacheKey(scopeRetrievalJob.Request.TenantId),
+                        new PaymentRequiredCacheEntry(ex.Message),
+                        "payment required state").ConfigureAwait(false);
+                }
+
+                break;
+        }
+    }
+
+    private async Task SetCacheBestEffortAsync<TKey, TValue>(TKey key, TValue value, string cacheEntryName)
```
Comment (Rishabh4275) [inline dotnet/src/Microsoft.Agents.AI.Purview/BackgroundJobRunner.cs]: There is no reason to have this.

File: dotnet/src/Microsoft.Agents.AI.Purview/README.md [hunk 8]
```diff
@@ -218,8 +218,12 @@ The policy logic is identical; the only difference is the hook point in the pipe
 
 The user id from the prompt message(s) is reused for the response evaluation so both evaluations map consistently to the same user.
 
-There are several optimizations to speed up Purview calls. Protection scope lookups (the first step in evaluation) are cached to minimize network calls. 
-If the policies allow content to be processed offline, the middleware will add the process content request to a channel and run it in a background worker. Similarly, the middleware will run a background request if no scopes apply and the interaction only has to be logged in Audit.
+There are several optimizations to speed up Purview calls:
```
Comment (Rishabh4275) [inline dotnet/src/Microsoft.Agents.AI.Purview/README.md]: Let's keep the old section itself. 
With readmes' it generally good to ask the agent to keep it consistent with the present writing style and across other READMEs in the project.

File: dotnet/src/Microsoft.Agents.AI.Purview/ScopedContentProcessor.cs [hunk 9]
```diff
@@ -193,43 +193,65 @@ private async Task<ProcessContentResponse> ProcessContentWithProtectionScopesAsy
     {
         ProtectionScopesRequest psRequest = CreateProtectionScopesRequest(pcRequest, pcRequest.UserId, pcRequest.TenantId, pcRequest.CorrelationId);
 
+        // Like the SDK tenant-scope sentinel cache, a 402 discovered by background
```
Comment (Rishabh4275) [inline dotnet/src/Microsoft.Agents.AI.Purview/ScopedContentProcessor.cs]: [nit] Let's only keep useful log statements, valid across the whole PR

File: python/packages/a2a/agent_framework_a2a/_agent.py [hunk 10]
```diff
@@ -157,9 +157,7 @@ def __init__(
             self.client = factory.create(agent_card, interceptors=interceptors)  # type: ignore
         except Exception as transport_error:
             # Transport negotiation failed - fall back to minimal agent card with JSONRPC
-            fallback_url = (
-                agent_card.supported_interfaces[0].url if agent_card.supported_interfaces else url
-            )
+            fallback_url = agent_card.supported_interfaces[0].url if agent_card.supported_interfaces else url
```
Comment (Rishabh4275) [inline python/packages/a2a/agent_framework_a2a/_agent.py]: Please don't make changes outside our packages if it's not a logic change that is actually needed.

File: python/packages/purview/AGENTS.md [hunk 11]
```diff
@@ -9,6 +9,13 @@ Integration with Microsoft Purview for data governance and policy enforcement.
 - **`PurviewPolicyMiddleware`** - Agent middleware for Purview policy enforcement
 - **`PurviewChatPolicyMiddleware`** - Chat-level middleware for policy enforcement
 
+The middleware implements a Purview SDK-aligned protection scopes flow:
```
Comment (Rishabh4275) [inline python/packages/purview/AGENTS.md]: Let's remove this, as we don't need to implementation details

File: dotnet/src/Microsoft.Agents.AI.Purview/BackgroundJobRunner.cs (hunk 7 above)
Comment (taisirhassan) [reply dotnet/src/Microsoft.Agents.AI.Purview/BackgroundJobRunner.cs]: inlined this call

File: dotnet/src/Microsoft.Agents.AI.Purview/ScopedContentProcessor.cs (hunk 9 above)
Comment (taisirhassan) [reply dotnet/src/Microsoft.Agents.AI.Purview/ScopedContentProcessor.cs]: removed

File: python/packages/a2a/agent_framework_a2a/_agent.py (hunk 10 above)
Comment (taisirhassan) [reply python/packages/a2a/agent_framework_a2a/_agent.py]: accidentally pulled in through formatter, removed changes

**Review by eavanvalkenburg — APPROVED** [review]

**Review by Rishabh4275 — APPROVED** [review]

**Review by eoindoherty1 — APPROVED** [review]
